### ISSUE TITLE
Enable config migration by default

### ISSUE THREAD

**Issue by HonkingGoose** [renovatebot/renovate#16359]
### What would you like Renovate to be able to do?

Right now users can opt-in to get config migration PRs by setting `configMigration=true`. I propose we make getting config migration PRs the default behavior for all users.

### Benefits:

- Config files match documentation, which prevents confusion
- People learn about the renamed features at once, when the config migration PR lands

### Drawbacks:

- Once we switch the default setting, Renovate creates many config migration PRs, which will be perceived as noise by our users

### Opting out by users

Users can opt-out of this new feature by setting  `configMigration=false` in their repository's `renovate.json`. Organizations can opt-out via their global `.github/renovate` repository.

### If you have any ideas on how this should be implemented, please tell us here.

Maybe add `configMigration=true` to our `config:base` and `config:best-practice` presets? Or maybe you want to hard-code this in deeper in the code?

### Is this a feature you are interested in implementing yourself?

No

**Comment by HonkingGoose** [comment]
I think we should put this change in a new major version of Renovate, so the changelog shows this as a breaking change in behavior. If you disagree, feel free to remove the label. 😉

**Comment by RahulGautamSingh** [comment]
Found a bug related to migration:
We delete comments from `renovate.json5` when migrating.
Reproduction: https://github.com/RahulGautamSingh-testing/renovate_json5/pull/4/files

**Comment by HonkingGoose** [comment]
@rarkins Are we `status:ready` on this issue? Or are there issues/things which really need to be done before we can make config migration the default behavior for all users?

**Comment by rarkins** [comment]
I think we should first try to get more use of this feature by prompting people in the dependency dashboard that to open the migration PR "on demand": #19783

**Comment by voxpelli** [comment]
Does this feature work for shared renovate configs that are split into multiple files? Like eg mine: https://github.com/voxpelli/renovate-config

**Comment by viceice** [comment]
no, currently only for the repo local config

**Comment by TWiStErRob** [comment]
Is there an issue tracking implementation of shared config repositories so we can upvote and subscribe to it?

**Comment by rarkins** [comment]
I couldn't find one, but created #22525

**Comment by rarkins** [comment]
I don't think this is ready until we can reduce the amount of unnecessary formatting changes, e.g.

https://github.com/renovate-reproductions/migrate1/pull/2/files

The amount of green/red makes it hard to see the _actual_ changes

**Comment by viceice** [comment]
> The good news is that adding a `.prettierrc` to that repository resulted in those changes mostly disappearing.
> 
> Bad news is that doesn't fix for non-prettier cases

fixing that without knowledge about the current formatting (like prettier) is pretty much impossible.

**Comment by voxpelli** [comment]
> fixing that without knowledge about the current formatting (like prettier) is pretty much impossible.

You would need to parse the JSON while persisting formatting data, you could maybe use [`tree-sitter-json`](https://github.com/tree-sitter/tree-sitter-json) for that or do some quick parsing yourself based on eg. some peg-style grammars out there, considering that JSOn is quite simple to parse. I did a quick rough incomplete such sketch here: https://gist.github.com/voxpelli/3af3a377ed13c2ffa7e507437974fd2b

**Comment by rarkins** [comment]
Another idea: parse the git diff output, then compare chunk by chunk. Parse each removed/added chunk and if they are equal then revert that change. A challenge would be where multiple changes are adjacent because then git groups them together.

**Comment by TWiStErRob** [comment]
> I don't think this is ready until we can reduce the amount of unnecessary formatting changes

I would think that's acceptable on the first enable of the migration, and after that the formatting and migrations would be kept up to date by smaller trivial PRs.

---

That said, I would love to option to keep my JSON formatting style and if adding a config for it is the way, then be it, but I wouldn't want it to affect JSON files other than than Renovate config, if that's possible.

I like @voxpelli's approach, clean and targeted. Keep in mind JSON5 and other formats though.

**Comment by TWiStErRob** [comment]
@rarkins would adding the prettierrc temporarily in the clone, without actually committing it, work?

**Comment by voxpelli** [comment]
There's this parser from @ota-meshi that supports JSON5, JSONC and plain JSON: https://github.com/ota-meshi/jsonc-eslint-parser

Though it seems to currently only do parsing, not serializing.

Ideally there would be a JSON Patch library built on top of such a white-space preserved AST and then a serialization of that AST.

**Comment by voxpelli** [comment]
Sorry for the spam here, but I did some more searching and found this module by @noahsug which seems to do the trick: https://www.npmjs.com/package/json5-writer

It didn't like `null` values, but apart from that, inserting the raw text of the source JSON and the parsed target JSON in the `.write()` seems to pretty much give the desired outcome in `.toSource()`

Should be possible to make use of without that many adaptions either.

**Comment by rarkins** [comment]
@voxpelli I like the idea of that library, but it's got some major challenges before we could adopt it:
- Low adoption
- Not maintained
- 0.x version
- Requires any changes to the AST to be done via `writer.write()` whereas we have written our migration to work on regular JS objects. We'd either need to change how we do migrations or add some type of mediation layer where we diff the final object and then apply changes to the writer

**Comment by voxpelli** [comment]
@rarkins My belief is that you can give the full modified JS object to `write()`

But yeah, would probably need to be forked and self-published unless the original author shows up.

**Comment by rarkins** [comment]
Adapting their example in this repo:

```
const fs = require('fs');
const json5Writer = require('json5-writer');

const config = fs.readFileSync('renovate.json', 'utf-8');
const writer = json5Writer.load(config);
writer.write({
  'eat honey': { cooldown: 3 },
  speak: { cooldown: 2 },
  bear: { actions: ['eat honey', 'speak'] },
});
fs.writeFileSync('renovate.json', writer.toSource(), 'utf-8');
```

It resulted in the renovate.json being fully wiped and only the new data remaining. If simply reading and writing it (i.e. dropping the `writer.write()` command then it's preserved.

**Comment by voxpelli** [comment]
> `.write(value)`
>
> Updates the JSON / JSON5 string with the new value. Any field or property that doesn't exist in value is removed.
>
> To keep an existing value, use `undefined`

@rarkins Either providing values for all keys, or values for all updates keys and setting the rest as `undefined`, would work

**Comment by voxpelli** [comment]
> So is their example in the README flawed? Why read in a file if you'll just blindly override it.

If you look at the diff after the example code you can see that their `config.json5` contains two of those fields.

But yes, one can argue that the example could be better and likewise the API, but it could be a good start 😌

I still think the ideal solution would be a JSON Patch library built on top of a white-space preserving AST they can then be serialized, but I also think that could be way out of scope for you to build here and that module was the closest my googling skills could find.

**Comment by reitzig** [comment]
FWIW: I for one am far less concerned with unwanted but easily remedied formatting changes (assuming my project is set up with linters and formatters to begin with wink 😉) than I am with my devs _getting_ config migration proposals.

**Comment by rarkins** [comment]
Let's add this in the next major (v38)

**Comment by HonkingGoose** [comment]
@rarkins earlier you said you wanted issue #19783 done _before_ making config migration the default:

> I think we should first try to get more use of this feature by prompting people in the dependency dashboard that to open the migration PR "on demand": #19783

I guess #19783 is _not blocking_ this issue anymore, because you removed the `status:blocked` label on this issue?

Just checking if we're really good to go to make config migration the default behavior soon. :smile:

**Comment by rarkins** [comment]
Yes, I think we can

**Comment by gberche-orange** [comment]
I also notice that Config migration PRs removes any existing Json 5 comments in the existing config file, see https://github.com/orange-cloudfoundry/k3s-packages-boshrelease/pull/87/files#diff-31b1aa2b93afe11792d84c6cb5057a064b5a3294fa0caab92d7570ea604f759f

This therefore requires manual merge of this change to restore json5 comments

**Comment by rarkins** [comment]
Marking this as blocked as should enable opt-in via dashboard first

**Comment by TheKevJames** [comment]
Semi-related idea / potential item to consider before making this default:

(not sure this is just a docs/comment update or a behaviour change, so including it in this thread for discussion before turning it into a real issue)

I've just enabled config migration and noticed some confusing wording leading me to be unsure whether my actions have disabled config migration in the future or not. This relates to https://github.com/renovatebot/renovate/issues/16359#issuecomment-1892594964 and eg. items such as #16383:

I use json5 configs, and the automated config migration removed comments / changed whitespace in ways I didn't prefer. As such, I made only the subset of config migration changes myself, then closed the automated PR. The automated PR, though, includes "🔕 Ignore: Close this PR and you won't be reminded about config migration again, but one day your current config may no longer be valid.". It's no clear as to whether this applies even though I applied the (relevant) subset of the change.

* [example automated PR](https://github.com/TheKevJames/tools/pull/1266)
* [my manual commit](https://github.com/TheKevJames/tools/commit/9c7adaf89dffc9886471647d41e7d50ae0810511)

**Comment by rarkins** [comment]
@TheKevJames I'm not sure but it's possible that you won't get any future migration PRs again due to that closing.

If already done, rename that PR to 'cache break' it.

In future, better that you keep the automated PR open until you've merged your own manual changes, I think it should then get autoclosed.

**Comment by TheKevJames** [comment]
Yup, seems doing the rename has avoided any issues -- purposely introducing something migrateable created a new PR. That probably means my comment can boil down to a doc issue, eg. including "Rename the closed PR to re-enable automated config migration." in the PR description might be enough to lead new users to the solution.
